File: research/system_research.py
import subprocess
import re

# research state management
state = {
    "is_mocked": False,
    "mock_values": {}
}

def run_command(cmd):
    """helper to execute shell commands and return clean output"""
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, shell=True)
        return result.stdout.strip()
    except Exception:
        return None

def get_native_hwid():
    """retrieves the actual hardware uuid from windows"""
    raw = run_command("wmic csproduct get uuid")
    if raw:
        # find uuid pattern using regex
        match = re.search(r'[A-F0-9]{8}-[A-F0-9]{4}-[A-F0-9]{4}-[A-F0-9]{4}-[A-F0-9]{12}', raw, re.I)
        if match:
            return match.group(0)
    return "unknown-hwid"

def get_native_mac():
    """retrieves the primary network adapter physical address"""
    raw = run_command("getmac /fo list /v")
    if raw:
        match = re.search(r'Physical Address:\s+([0-9A-F-]+)', raw, re.I)
        if match:
            return match.group(1)
    return "00-00-00-00-00-00"

def get_hardware_id():
    """the function targeted by research (hookable)"""
    if state["is_mocked"] and "hwid" in state["mock_values"]:
        return state["mock_values"]["hwid"]
    return get_native_hwid()

def get_mac_address():
    """the function targeted by research (hookable)"""
    if state["is_mocked"] and "mac" in state["mock_values"]:
        return state["mock_values"]["mac"]
    return get_native_mac()

def execute_security_audit():
    """simulates an application performing a hardware-bound security check"""
    current_hwid = get_hardware_id()
    current_mac = get_mac_address()
    
    print(f" [audit] hwid detected: {current_hwid}")
    print(f" [audit] mac detected:  {current_mac}")
    
    # example research whitelist
    whitelist = ["550e8400-e29b-41d4-a716-446655440000"]
    
    if current_hwid in whitelist:
        print(" [result] environment: authorized (virtual/mock match)")
    else:
        print(" [result] environment: native/unrecognized")

def main():
    print("="*60)
    print(" python system identifier research & simulation tool")
    print("="*60)
    
    # phase 1: native state
    print("\n[step 1] analyzing native host identifiers (before mock)")
    execute_security_audit()
    
    # phase 2: mock configuration
    print("\n[step 2] initializing virtualization simulation layer...")
    state["mock_values"]["hwid"] = "550e8400-e29b-41d4-a716-446655440000"
    state["mock_values"]["mac"] = "DE-AD-BE-EF-CA-FE"
    state["is_mocked"] = True
    print(" [system] mock layer activated.")
    
    # phase 3: mocked state
    print("\n[step 3] analyzing simulated identifiers (after mock)")
    execute_security_audit()
    
    print("\n" + "="*60)

if __name__ == "__main__":
    main()

File: research/system_research.lua
-- system_research.lua
-- educational script for exploring system identifier detection and simulation
-- purpose: virtualization research and software testing environments

local state = {
    cache = {},
    is_mocked = false,
    mock_values = {}
}

-- helper: execute system commands via shell
local function run_shell(cmd)
    local pipe = io.popen(cmd)
    if not pipe then return nil end
    local output = pipe:read("*a")
    pipe:close()
    return output:match("^%s*(.-)%s*$")
end

-- retrieve native hardware uuid
local function get_native_hwid()
    local raw = run_shell("wmic csproduct get uuid")
    if raw then
        return raw:match("%w%w%w%w%w%w%w%w%-%w%w%w%w%-%w%w%w%w%-%w%w%w%w%-%w%w%w%w%w%w%w%w%w%w%w%w") or "unknown"
    end
    return "error"
end

-- retrieve native mac address
local function get_native_mac()
    local raw = run_shell("getmac /fo list /v")
    if raw then
        return raw:match("Physical Address:%s+([%w%-]+)") or "00-00-00-00-00-00"
    end
    return "00-00-00-00-00-00"
end

-- core detection functions (can be hooked for research)
local function get_hardware_id()
    return get_native_hwid()
end

local function get_mac_address()
    return get_native_mac()
end

-- research: function hooking demonstration
-- this simulates how a virtualization layer or 'spoof' might intercept calls
local function apply_research_hooks()
    print("\n[research] applying api hooks for identifier redirection...")
    
    -- store originals
    local original_hwid = get_hardware_id
    local original_mac  = get_mac_address
    
    -- redirect hardware id
    get_hardware_id = function()
        if state.is_mocked then
            return state.mock_values.hwid
        end
        return original_hwid()
    end
    
    -- redirect mac address
    get_mac_address = function()
        if state.is_mocked then
            return state.mock_values.mac
        end
        return original_mac()
    end
    
    print("[research] hooks installed successfully.")
end

-- simulation: application-level security check
local function execute_identity_verification()
    print("\n[app] starting identity verification...")
    
    local hwid = get_hardware_id()
    local mac  = get_mac_address()
    
    print(string.format(" [app] detected hwid: %s", hwid))
    print(string.format(" [app] detected mac:  %s", mac))
    
    -- demo whitelist
    local whitelist = {
        ["550e8400-e29b-41d4-a716-446655440000"] = true
    }
    
    if whitelist[hwid] then
        print("[app] result: authorized.")
    else
        print("[app] result: unauthorized device.")
    end
end

-- entry point
local function main()
    print("====================================================")
    print(" system identifier research & simulation tool")
    print("====================================================")
    
    -- 1. show native data
    print("\n--- step 1: native environment analysis ---")
    execute_identity_verification()
    
    -- 2. apply hooks (research technique)
    apply_research_hooks()
    
    -- 3. activate mock layer
    state.mock_values.hwid = "550e8400-e29b-41d4-a716-446655440000"
    state.mock_values.mac  = "DE-AD-BE-EF-CA-FE"
    state.is_mocked = true
    
    -- 4. test intercepted environment
    print("\n--- step 2: simulated environment analysis ---")
    execute_identity_verification()
    
    print("\n====================================================")
end

main()
